File: app/lib/versification.ts
/**
 * Tables de correspondance versification catholique ↔ protestante.
 *
 * La numérotation de référence dans cette base est CATHOLIQUE/VULGATE.
 * Ce module convertit les références protestantes (hébreu/LXX anglicane)
 * en catholiques pour l'import, et vice-versa pour l'export.
 *
 * Livres concernés : PSA, EXO, LEV, NUM, DEU, HOS, JOL, MAL
 */

export type Ref = { ch: number; v: number };

// ─── Psaumes ────────────────────────────────────────────────────────────────
// Règle générale : les psaumes 1-8 et 148-150 sont identiques dans les deux
// traditions. Entre les deux, la Vulgate fusionne Ps 9 et 10 hébreux → Ps 9
// catholique (39v), puis décale de -1 jusqu'à Ps 112 cath = Ps 113 héb,
// avant de fusionner à nouveau Ps 113 héb en 114+115 cath, etc.
//
// Tables complètes de la divergence (sens cath → prot) :
//   Cath 9:1-21    → Prot 9:1-21
//   Cath 9:22-39   → Prot 10:1-18   (v_prot = v_cath - 21)
//   Cath 10-112    → Prot ch+1      (même verset)
//   Cath 113:1-8   → Prot 114:1-8
//   Cath 113:9-26  → Prot 115:1-18  (v_prot = v_cath - 8)
//   Cath 114:1-9   → Prot 116:1-9
//   Cath 115:1-10  → Prot 116:10-19 (v_prot = v_cath + 9)
//   Cath 116-145   → Prot ch+1      (même verset)
//   Cath 146:1-11  → Prot 147:1-11
//   Cath 147:1-20  → Prot 147:12-31 (v_prot = v_cath + 11)
//   Cath 148-150   → Prot 148-150

export function psaCathVerseProt(ch: number, v: number): Ref {
  if (ch <= 8) return { ch, v };

  if (ch === 9) {
    if (v <= 21) return { ch: 9, v };
    return { ch: 10, v: v - 21 };
  }

  if (ch >= 10 && ch <= 112) return { ch: ch + 1, v };

  if (ch === 113) {
    if (v <= 8) return { ch: 114, v };
    return { ch: 115, v: v - 8 };
  }

  if (ch === 114) return { ch: 116, v };
  if (ch === 115) return { ch: 116, v: v + 9 };

  if (ch >= 116 && ch <= 145) return { ch: ch + 1, v };

  if (ch === 146) return { ch: 147, v };
  if (ch === 147) return { ch: 147, v: v + 11 };

  // 148-150
  return { ch, v };
}

export function psaProtVerseCath(ch: number, v: number): Ref {
  if (ch <= 8) return { ch, v };

  if (ch === 9) return { ch: 9, v };
  if (ch === 10) return { ch: 9, v: v + 21 };

  if (ch >= 11 && ch <= 113) return { ch: ch - 1, v };

  if (ch === 114) return { ch: 113, v };
  if (ch === 115) return { ch: 113, v: v + 8 };
  if (ch === 116) {
    if (v <= 9) return { ch: 114, v };
    return { ch: 115, v: v - 9 };
  }

  if (ch >= 117 && ch <= 146) return { ch: ch - 1, v };

  if (ch === 147) {
    if (v <= 11) return { ch: 146, v };
    return { ch: 147, v: v - 11 };
  }

  // 148-150
  return { ch, v };
}

// ─── Exode ──────────────────────────────────────────────────────────────────
// Cath 8:1-4   → Prot 7:26-29   (ch=7, v_prot = v_cath + 25)
// Cath 8:5+    → Prot 8:1+      (v_prot = v_cath - 4)
// Cath 7:1-25  → Prot 7:1-25    (identique)

export function exoCathVerseProt(ch: number, v: number): Ref {
  if (ch !== 8) return { ch, v };
  if (v <= 4) return { ch: 7, v: v + 25 };
  return { ch: 8, v: v - 4 };
}

export function exoProtVerseCath(ch: number, v: number): Ref {
  if (ch === 7 && v >= 26) return { ch: 8, v: v - 25 };
  if (ch === 8) return { ch: 8, v: v + 4 };
  return { ch, v };
}

// ─── Lévitique ──────────────────────────────────────────────────────────────
// Cath 5:20-26 → Prot 6:1-7    (ch=6, v_prot = v_cath - 19)
// Cath 6:1-23  → Prot 6:8-30   (ch=6, v_prot = v_cath + 7)
// Tout le reste identique.

export function levCathVerseProt(ch: number, v: number): Ref {
  if (ch === 5 && v >= 20) return { ch: 6, v: v - 19 };
  if (ch === 6) return { ch: 6, v: v + 7 };
  return { ch, v };
}

export function levProtVerseCath(ch: number, v: number): Ref {
  if (ch === 6) {
    if (v <= 7) return { ch: 5, v: v + 19 };
    return { ch: 6, v: v - 7 };
  }
  return { ch, v };
}

// ─── Nombres ────────────────────────────────────────────────────────────────
// Cath 17:1-15  → Prot 16:36-50  (ch=16, v_prot = v_cath + 35)
// Cath 17:16-28 → Prot 17:1-13   (ch=17, v_prot = v_cath - 15)
// Cath 16:1-35  → Prot 16:1-35   (identique)

export function numCathVerseProt(ch: number, v: number): Ref {
  if (ch === 17) {
    if (v <= 15) return { ch: 16, v: v + 35 };
    return { ch: 17, v: v - 15 };
  }
  return { ch, v };
}

export function numProtVerseCath(ch: number, v: number): Ref {
  if (ch === 16 && v >= 36) return { ch: 17, v: v - 35 };
  if (ch === 17) return { ch: 17, v: v + 15 };
  return { ch, v };
}

// ─── Deutéronome ────────────────────────────────────────────────────────────
// Cath 13:1    → Prot 12:32   (ch=12, v=32)
// Cath 13:2-19 → Prot 13:1-18 (ch=13, v_prot = v_cath - 1)
// Cath 12:1-31 → Prot 12:1-31 (identique)

export function deuCathVerseProt(ch: number, v: number): Ref {
  if (ch === 13) {
    if (v === 1) return { ch: 12, v: 32 };
    return { ch: 13, v: v - 1 };
  }
  return { ch, v };
}

export function deuProtVerseCath(ch: number, v: number): Ref {
  if (ch === 12 && v === 32) return { ch: 13, v: 1 };
  if (ch === 13) return { ch: 13, v: v + 1 };
  return { ch, v };
}

// ─── Osée ───────────────────────────────────────────────────────────────────
// Cath 2:1-2   → Prot 1:10-11  (ch=1, v_prot = v_cath + 9)
// Cath 2:3-25  → Prot 2:1-23   (ch=2, v_prot = v_cath - 2)
// Cath 1:1-9   → Prot 1:1-9    (identique)

export function hosCathVerseProt(ch: number, v: number): Ref {
  if (ch === 2) {
    if (v <= 2) return { ch: 1, v: v + 9 };
    return { ch: 2, v: v - 2 };
  }
  return { ch, v };
}

export function hosProtVerseCath(ch: number, v: number): Ref {
  if (ch === 1 && v >= 10) return { ch: 2, v: v - 9 };
  if (ch === 2) return { ch: 2, v: v + 2 };
  return { ch, v };
}

// ─── Joël ───────────────────────────────────────────────────────────────────
// Cath 3:1-5  → Prot 2:28-32   (ch=2, v_prot = v_cath + 27)
// Cath 4:1-21 → Prot 3:1-21    (ch=3, même verset)
// Cath 1-2:27 → Prot 1-2:27    (identique)

export function jolCathVerseProt(ch: number, v: number): Ref {
  if (ch === 3) return { ch: 2, v: v + 27 };
  if (ch === 4) return { ch: 3, v };
  return { ch, v };
}

export function jolProtVerseCath(ch: number, v: number): Ref {
  if (ch === 2 && v >= 28) return { ch: 3, v: v - 27 };
  if (ch === 3) return { ch: 4, v };
  return { ch, v };
}

// ─── Malachie ───────────────────────────────────────────────────────────────
// Cath 4:1-6  → Prot 3:19-24   (ch=3, v_prot = v_cath + 18)
// Cath 1-3:18 → Prot 1-3:18    (identique)

export function malCathVerseProt(ch: number, v: number): Ref {
  if (ch === 4) return { ch: 3, v: v + 18 };
  return { ch, v };
}

export function malProtVerseCath(ch: number, v: number): Ref {
  if (ch === 3 && v >= 19) return { ch: 4, v: v - 18 };
  return { ch, v };
}

// ─── Fonction générale ───────────────────────────────────────────────────────

type Livre = "PSA" | "EXO" | "LEV" | "NUM" | "DEU" | "HOS" | "JOL" | "MAL";

const CONVERTISSEURS_CATH_PROT: Partial<
  Record<string, (ch: number, v: number) => Ref>
> = {
  PSA: psaCathVerseProt,
  EXO: exoCathVerseProt,
  LEV: levCathVerseProt,
  NUM: numCathVerseProt,
  DEU: deuCathVerseProt,
  HOS: hosCathVerseProt,
  JOL: jolCathVerseProt,
  MAL: malCathVerseProt,
};

const CONVERTISSEURS_PROT_CATH: Partial<
  Record<string, (ch: number, v: number) => Ref>
> = {
  PSA: psaProtVerseCath,
  EXO: exoProtVerseCath,
  LEV: levProtVerseCath,
  NUM: numProtVerseCath,
  DEU: deuProtVerseCath,
  HOS: hosProtVerseCath,
  JOL: jolProtVerseCath,
  MAL: malProtVerseCath,
};

/** Convertit une référence catholique en référence protestante. */
export function cathVerseProt(
  livre: string,
  ch: number,
  v: number
): Ref {
  const fn = CONVERTISSEURS_CATH_PROT[livre];
  return fn ? fn(ch, v) : { ch, v };
}

/** Convertit une référence protestante en référence catholique. */
export function protVerseCath(
  livre: string,
  ch: number,
  v: number
): Ref {
  const fn = CONVERTISSEURS_PROT_CATH[livre];
  return fn ? fn(ch, v) : { ch, v };
}

/** Retourne vrai si le livre a des divergences de numérotation. */
export function aDesConvergences(livre: string): boolean {
  return livre in CONVERTISSEURS_CATH_PROT;
}

File: scripts/versification.py
"""
Tables de correspondance versification catholique ↔ protestante.

La numérotation de référence dans cette base est CATHOLIQUE/VULGATE.
Ce module convertit les références protestantes en catholiques pour
l'import, et vice-versa pour l'export.

Livres concernés : PSA, EXO, LEV, NUM, DEU, HOS, JOL, MAL

Usage:
    from versification import cath_vers_prot, prot_vers_cath

    ch_prot, v_prot = cath_vers_prot("PSA", 9, 25)   # → (10, 4)
    ch_cath, v_cath = prot_vers_cath("PSA", 10, 4)    # → (9, 25)
"""


def psa_cath_vers_prot(ch, v):
    if ch <= 8:
        return ch, v
    if ch == 9:
        return (9, v) if v <= 21 else (10, v - 21)
    if 10 <= ch <= 112:
        return ch + 1, v
    if ch == 113:
        return (114, v) if v <= 8 else (115, v - 8)
    if ch == 114:
        return 116, v
    if ch == 115:
        return 116, v + 9
    if 116 <= ch <= 145:
        return ch + 1, v
    if ch == 146:
        return 147, v
    if ch == 147:
        return 147, v + 11
    return ch, v  # 148-150


def psa_prot_vers_cath(ch, v):
    if ch <= 8:
        return ch, v
    if ch == 9:
        return 9, v
    if ch == 10:
        return 9, v + 21
    if 11 <= ch <= 113:
        return ch - 1, v
    if ch == 114:
        return 113, v
    if ch == 115:
        return 113, v + 8
    if ch == 116:
        return (114, v) if v <= 9 else (115, v - 9)
    if 117 <= ch <= 146:
        return ch - 1, v
    if ch == 147:
        return (146, v) if v <= 11 else (147, v - 11)
    return ch, v  # 148-150


def exo_cath_vers_prot(ch, v):
    if ch != 8:
        return ch, v
    if v <= 4:
        return 7, v + 25
    return 8, v - 4


def exo_prot_vers_cath(ch, v):
    if ch == 7 and v >= 26:
        return 8, v - 25
    if ch == 8:
        return 8, v + 4
    return ch, v


def lev_cath_vers_prot(ch, v):
    if ch == 5 and v >= 20:
        return 6, v - 19
    if ch == 6:
        return 6, v + 7
    return ch, v


def lev_prot_vers_cath(ch, v):
    if ch == 6:
        return (5, v + 19) if v <= 7 else (6, v - 7)
    return ch, v


def num_cath_vers_prot(ch, v):
    if ch == 17:
        return (16, v + 35) if v <= 15 else (17, v - 15)
    return ch, v


def num_prot_vers_cath(ch, v):
    if ch == 16 and v >= 36:
        return 17, v - 35
    if ch == 17:
        return 17, v + 15
    return ch, v


def deu_cath_vers_prot(ch, v):
    if ch == 13:
        return (12, 32) if v == 1 else (13, v - 1)
    return ch, v


def deu_prot_vers_cath(ch, v):
    if ch == 12 and v == 32:
        return 13, 1
    if ch == 13:
        return 13, v + 1
    return ch, v


def hos_cath_vers_prot(ch, v):
    if ch == 2:
        return (1, v + 9) if v <= 2 else (2, v - 2)
    return ch, v


def hos_prot_vers_cath(ch, v):
    if ch == 1 and v >= 10:
        return 2, v - 9
    if ch == 2:
        return 2, v + 2
    return ch, v


def jol_cath_vers_prot(ch, v):
    if ch == 3:
        return 2, v + 27
    if ch == 4:
        return 3, v
    return ch, v


def jol_prot_vers_cath(ch, v):
    if ch == 2 and v >= 28:
        return 3, v - 27
    if ch == 3:
        return 4, v
    return ch, v


def mal_cath_vers_prot(ch, v):
    if ch == 4:
        return 3, v + 18
    return ch, v


def mal_prot_vers_cath(ch, v):
    if ch == 3 and v >= 19:
        return 4, v - 18
    return ch, v


_CATH_PROT = {
    "PSA": psa_cath_vers_prot,
    "EXO": exo_cath_vers_prot,
    "LEV": lev_cath_vers_prot,
    "NUM": num_cath_vers_prot,
    "DEU": deu_cath_vers_prot,
    "HOS": hos_cath_vers_prot,
    "JOL": jol_cath_vers_prot,
    "MAL": mal_cath_vers_prot,
}

_PROT_CATH = {
    "PSA": psa_prot_vers_cath,
    "EXO": exo_prot_vers_cath,
    "LEV": lev_prot_vers_cath,
    "NUM": num_prot_vers_cath,
    "DEU": deu_prot_vers_cath,
    "HOS": hos_prot_vers_cath,
    "JOL": jol_prot_vers_cath,
    "MAL": mal_prot_vers_cath,
}


def cath_vers_prot(livre, ch, v):
    """Convertit une référence catholique en référence protestante.
    Retourne (ch_prot, v_prot). Identité si le livre ne diverge pas."""
    fn = _CATH_PROT.get(livre)
    return fn(ch, v) if fn else (ch, v)


def prot_vers_cath(livre, ch, v):
    """Convertit une référence protestante en référence catholique.
    Retourne (ch_cath, v_cath). Identité si le livre ne diverge pas."""
    fn = _PROT_CATH.get(livre)
    return fn(ch, v) if fn else (ch, v)


def a_des_divergences(livre):
    """Retourne True si le livre a des différences de numérotation."""
    return livre in _CATH_PROT


# ─── Tests rapides ───────────────────────────────────────────────────────────

if __name__ == "__main__":
    cas = [
        # (livre, sens, ch_in, v_in, ch_out_attendu, v_out_attendu)
        ("PSA", "c→p", 9, 1,   9, 1),
        ("PSA", "c→p", 9, 21,  9, 21),
        ("PSA", "c→p", 9, 22, 10, 1),
        ("PSA", "c→p", 9, 39, 10, 18),
        ("PSA", "c→p", 10, 1, 11, 1),
        ("PSA", "c→p", 112, 1, 113, 1),
        ("PSA", "c→p", 113, 1, 114, 1),
        ("PSA", "c→p", 113, 8, 114, 8),
        ("PSA", "c→p", 113, 9, 115, 1),
        ("PSA", "c→p", 113, 26, 115, 18),
        ("PSA", "c→p", 114, 1, 116, 1),
        ("PSA", "c→p", 115, 1, 116, 10),
        ("PSA", "c→p", 146, 5, 147, 5),
        ("PSA", "c→p", 147, 1, 147, 12),
        ("PSA", "c→p", 150, 6, 150, 6),
        ("EXO", "c→p", 8, 1, 7, 26),
        ("EXO", "c→p", 8, 4, 7, 29),
        ("EXO", "c→p", 8, 5, 8, 1),
        ("LEV", "c→p", 5, 20, 6, 1),
        ("LEV", "c→p", 6, 1, 6, 8),
        ("NUM", "c→p", 17, 1, 16, 36),
        ("NUM", "c→p", 17, 15, 16, 50),
        ("NUM", "c→p", 17, 16, 17, 1),
        ("DEU", "c→p", 13, 1, 12, 32),
        ("DEU", "c→p", 13, 2, 13, 1),
        ("HOS", "c→p", 2, 1, 1, 10),
        ("HOS", "c→p", 2, 2, 1, 11),
        ("HOS", "c→p", 2, 3, 2, 1),
        ("JOL", "c→p", 3, 1, 2, 28),
        ("JOL", "c→p", 4, 1, 3, 1),
        ("MAL", "c→p", 4, 1, 3, 19),
        ("MAL", "c→p", 4, 6, 3, 24),
    ]

    erreurs = 0
    for livre, sens, ch_in, v_in, ch_exp, v_exp in cas:
        if sens == "c→p":
            ch_out, v_out = cath_vers_prot(livre, ch_in, v_in)
        else:
            ch_out, v_out = prot_vers_cath(livre, ch_in, v_in)
        ok = ch_out == ch_exp and v_out == v_exp
        if not ok:
            print(f"ERREUR {livre} {sens} {ch_in}:{v_in} → {ch_out}:{v_out}  (attendu {ch_exp}:{v_exp})")
            erreurs += 1

    if erreurs == 0:
        print(f"OK — {len(cas)} cas vérifiés sans erreur.")
    else:
        print(f"{erreurs} erreur(s) sur {len(cas)} cas.")
